network-api/networkapi/wagtailpages/pagemodels/products.py

```diff
@@ -426,6 +510,73 @@ def get_context(self, request, *args, **kwargs):
         )} - {self.title}'''
         return context
 
+    def serve(self, request, *args, **kwargs):
+        # In Wagtail we use the serve() method to detect POST submissions.
+        # Alternatively, this could be a routable view.
+        # For more on this, see the docs here:
+        # https://docs.wagtail.io/en/stable/reference/pages/model_recipes.html#overriding-the-serve-method
+        if request.body and request.method == "POST":
+            # If the request is POST. Parse the body.
+            data = json.loads(request.body)
+            # If the POST body has a productID and value, it's someone voting on the product
+            if data.get('productID') and data.get("value"):
+                # Product ID and Value can both be zero. It's impossible to get a Page with ID of zero.
+                try:
+                    product_id = int(data['productID'])  # ie. 68
+                    value = int(data["value"])  # ie. 0 to 100
+                except ValueError:
+                    return HttpResponseNotAllowed('Product ID or value is invalid')
+
+                if value < 0 or value > 100:
+                    return HttpResponseNotAllowed('Cannot save vote')
+
+                try:
+                    product = ProductPage.objects.get(id=product_id)
+                    # If the product exists but isn't live and the user isn't logged in..
+                    if not product.live and not request.user.is_authenticated:
+                        return HttpResponseNotFound("Product does not exist")
```

Comment:
just realised: we forgot the `if product is None` case, because someone could script themselves a vote bot that just votes for products from 1 to 1 million, and that's going to cause a lot of ProductPageNotFound errors O_O

Reply:
I'm going to add this in for good measure. The line above it: `product = ProductPage.objects.get(id=product_id)` should throw an exception if the product isn't found by its product_id and caught on L561.